<!DOCTYPE HTML>
<html xmlns:th="http://www.thymeleaf.org"
	xmlns:layout="http://www.ultraq.net.nz/web/thymeleaf/layout"
	layout:decorator="layout">
<body>
	<h1 layout:fragment="header">Python Client:</h1>
	<div layout:fragment="content" class="container">
                <p><h3>Basic Python Program</h3></p>
                <p>The simple python template program provides a basic set of
                        functions for querying the G2S API via the Python
                        platform for scientific computing.</p>
                <pre>
#Python 2.7
#A Simple Python template program reading G2S APIs

import requests
from requests.auth import HTTPDigestAuth
import json

#Setup API, e.g. http://g2s.genomenexus.org/swagger-ui.html#!/Genome/getPdbResidueByEnsemblIdGenomeUsingGET
#Can change to any valid URL in G2S API

hostName="http://g2s.genomenexus.org/g2s/"
apiName="GenomeStructureResidueMapping"
chromsomeNum="X"
nuclPosition="66937331"
nuclType="T"

# Set up API URL
url = hostName+apiName+"/"+chromsomeNum+"/"+nuclPosition+"/"+nuclType

# Request API
myResponse = requests.get(url)
#print (myResponse.status_code)

# For successful API call, response code will be 200 (OK)
if(myResponse.ok):

    # Loads (Load String) takes a Json file and converts into python data structure (dict or list, depending on JSON)
        # In this Example, jData are lists of Residues from Genome Mapping to Protein Structures
    jData = json.loads(myResponse.content)

        #pring output
    print jData
    print "\n"
else:
  # If response code is not ok (200), print the resulting http error code with description
    myResponse.raise_for_status()
        </pre>
<h3><p>Comprehensive Python Client</p></h3>
<p>A comprehensive python client can be generated by swagger-Codegen as follows:</p>

        <pre>
# swagger_client
A Genome to Strucure (G2S) API Supports Automated Mapping and Annotating Genomic Variants in 3D Protein Structures. Supports Inputs from Human Genome Position, Uniprot and Human Ensembl Names

This Python package is automatically generated by the [Swagger Codegen](https://github.com/swagger-api/swagger-codegen) project:

- API version: 2.0
- Package version: 1.0.0
- Build package: class io.swagger.codegen.languages.PythonClientCodegen

## Requirements.

Python 2.7

## Installation and Usage

git clone https://github.com/swagger-api/swagger-codegen
cd swagger-codegen
mvn clean package
java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate \
-i http://g2s.genomenexus.org/v2/api-docs?group=g2s \
-l python \
-o {YOURFOLDER/}py_api_client

(if you're on Windows, replace the last command with java -jar modules\swagger-codegen-cli\target\swagger-codegen-cli.jar generate -i http://g2s.genomenexus.org/v2/api-docs?group=g2s  -l python
-o {YOURFOLDER\}py_api_client)



## edit the Default Base url

Change the file called configuration.py which in the folder called swagger_client

self.host = "https://g2s.genomenexus.org/" -> self.host = "http://g2s.genomenexus.org/"


### pip install (optional)

If you have not install certifi, you may need to run `pip install certif`
`

### Setuptools

Install via [Setuptools](http://pypi.python.org/pypi/setuptools).

```sh
python setup.py install --user
```
`

## Getting Started

Please follow the [installation procedure](#installation--usage) and then run the following:

```python
from __future__ import print_function
import time
import swagger_client
from swagger_client.rest import ApiException
from pprint import pprint
# create an instance of the API class
api_instance = swagger_client.GenomeApi()
chromosome_num = 'X' # str | Input Chromomsome Number e.g. X
position_num = 66937331 # int | Input Nucleotide Position e.g. 66937331
nucleotide_type = 'T' # str | Input Nucleotide Type e.g. T Note: Please Input Correct Type

try:
    # Whether Position Valid in Genome
    api_response = api_instance.get_existed_ensembl_idin_genome_using_get(chromosome_num, position, nucleotide_type)
    pprint(api_response)
except ApiException as e:
    print("Exception when calling GenomeApi->get_existed_ensembl_idin_genome_using_get: %s\n" % e)

```
`

## Documentation for API Endpoints

All URIs are relative to *https://g2s.genomenexus.org/*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*GenomeApi* | [**get_existed_ensembl_idin_genome_using_get**](docs/GenomeApi.md#get_existed_ensembl_idin_genome_using_get) | **GET** /g2s/GenomeStructureRecognition/{chromosomeNum}/{position}/{nucleotideType} | Whether Position Valid in Genome
*GenomeApi* | [**get_pdb_residue_by_ensembl_id_genome_using_get**](docs/GenomeApi.md#get_pdb_residue_by_ensembl_id_genome_using_get) | **GET** /g2s/GenomeStructureResidueMapping/{chromosomeNum}/{position}/{nucleotideType} | Get Residue Mapping from Genome Position
*HumanEnsemblApi* | [**get_existed_ensembl_idin_alignment_using_get**](docs/HumanEnsemblApi.md#get_existed_ensembl_idin_alignment_using_get) | **GET** /g2s/EnsemblRecognition/{ensemblId} | Whether EnsemblId exists
*HumanEnsemblApi* | [**get_pdb_alignment_by_ensembl_id_using_get**](docs/HumanEnsemblApi.md#get_pdb_alignment_by_ensembl_id_using_get) | **GET** /g2s/EnsemblStructureMapping/{ensemblId} | Get PDB Alignments by EnsemblId
*HumanEnsemblApi* | [**get_pdb_residue_by_ensembl_id_using_get**](docs/HumanEnsemblApi.md#get_pdb_residue_by_ensembl_id_using_get) | **GET** /g2s/EnsemblResidueMapping/{ensemblId}/{position} | Get Residue Mapping by EnsemblId and Residue Number
*ProteinSeqIdApi* | [**get_existed_seq_idin_alignment_using_get**](docs/ProteinSeqIdApi.md#get_existed_seq_idin_alignment_using_get) | **GET** /g2s/GeneSeqRecognition/{seqId} | Whether Protein SeqId Exists
*ProteinSeqIdApi* | [**get_pdb_alignment_by_gene_sequence_id_using_get**](docs/ProteinSeqIdApi.md#get_pdb_alignment_by_gene_sequence_id_using_get) | **GET** /g2s/GeneSeqStructureMapping/{seqId} | Get PDB Alignments by Protein SeqId
*ProteinSeqIdApi* | [**get_pdb_residue_by_seq_id_using_get**](docs/ProteinSeqIdApi.md#get_pdb_residue_by_seq_id_using_get) | **GET** /g2s/GeneSeqResidueMapping/{seqId}/{position} | Get Residue Mapping by Protein SeqId and Residue Position
*UniprotApi* | [**get_existed_uniprot_id_isoin_alignment_using_get**](docs/UniprotApi.md#get_existed_uniprot_id_isoin_alignment_using_get) | **GET** /g2s/UniprotIsoformRecognition/{uniprotId}/{isoform} | Whether Isoform of UniprotId Exists
*UniprotApi* | [**get_existed_uniprot_idin_alignment_using_get**](docs/UniprotApi.md#get_existed_uniprot_idin_alignment_using_get) | **GET** /g2s/UniprotRecognition/{uniprotId} | Whether UniprotId Exists
*UniprotApi* | [**get_pdb_alignment_by_uniprot_id_iso_using_get**](docs/UniprotApi.md#get_pdb_alignment_by_uniprot_id_iso_using_get) | **GET** /g2s/UniprotIsoformStructureMapping/{uniprotId}/{isoform} | Get PDB Alignments by UniprotId and Isofrom
*UniprotApi* | [**get_pdb_alignment_by_uniprot_id_using_get**](docs/UniprotApi.md#get_pdb_alignment_by_uniprot_id_using_get) | **GET** /g2s/UniprotStructureMapping/{uniprotId} | Get PDB Alignments by UniprotId
*UniprotApi* | [**get_pdb_residue_by_uniprot_id_iso_using_get**](docs/UniprotApi.md#get_pdb_residue_by_uniprot_id_iso_using_get) | **GET** /g2s/UniprotIsoformResidueMapping/{uniprotId}/{isoform}/{position} | Get Residue Mapping by UniprotId, Isofrom and Residue Position
*UniprotApi* | [**get_pdb_residue_by_uniprot_id_using_get**](docs/UniprotApi.md#get_pdb_residue_by_uniprot_id_using_get) | **GET** /g2s/UniprotResidueMapping/{uniprotId}/{position} | Get Residue Mapping by UniprotId and Residue Position



## Documentation For Models

 - [Alignment](docs/Alignment.md)
 - [GenomeResidue](docs/GenomeResidue.md)
 - [Residue](docs/Residue.md)


## Documentation For Authorization

 All endpoints do not require authorization.


## Author
        </pre>
	</div>
</body>
</html>